README.md
## IPK Sniffer

The IPK sniffer is a C-based network sniffer which is inspired by [wireshar](https://www.wireshark.org/) sniffer, but it is a lot more simple to use. It can capture and analyze network IPv4, IPv6 and ARP packets of various types on a specified network interface.

### Installation
> **Note!**
The program is made for UNIX-based systems only. In other cases the sniffer won't work.

To install the sniffer program, follow these steps:
1. Download the source code from the repository.
2. Compile the source code using Makefile:
```bash
make
```
3. Run the program (launch example)
```bash
./ipk-sniffer -i eth0
```

- `ipk-sniffer` is the name of the sniffer program that will be created after `make`.
- `-i` option stands for **interface** sniffer captures packets from.

Program will start and print out first packet that it captures.

### Usage

```bash
./ipk-sniffer [-i interface | --interface interface] {-p port} {[--tcp|-t] [--udp|-u] [--arp] [--icmp] } {-n num}
```

- `-i interface` or `--interface interface`, where **interface** is network interface sniffer will listen to. If **interface** is not specified then the program will print list of available interfaces on user's device.
- `-p port` will filter captured packets by port. If this option is not set - all ports are considered.
`-t` or `--tcp` will show only TCP packets.
`-u` or `--idp` will show only UDP packets.
`--icmp4` will show only ICMPv4 packets.
`--icmp6` will display only ICMPv6 echo request/response.
`--arp` will display only ARP frames.
`--ndp` will display only ICMPv6 NDP packets.
`--igmp` will display only IGMP packets
`--mld` will display only MLD packets.
If none protocols are specified for filtering - all protocols are considered.
`-n N` where (N >= 0). It limits the amout of packets that should be captured by sniffer. By defailt its value is 1 (one packet).
### Output
Capture packet will have the following form:
```bash
timestamp: 2021-03-19T18:42:52.362+01:00
src MAC: 00:1c:2e:92:03:80
dst MAC: 00:1b:3f:56:8a:00
frame length: 512 bytes
src IP: 147.229.13.223
dst IP: 10.10.10.56
src port: 4093
dst port: 80

0x0000: 00 19 d1 f7 be e5 00 04 96 1d 34 20 08 00 45 00 ........ ..4 ..
0x0010: 05 a0 52 5b 40 00 36 06 5b db d9 43 16 8c 93 e5 ..R[@.6. [..C....
0x0020: 0d 6d 00 50 0d fb 3d cd 0a ed 41 d1 a4 ff 50 18 .m.P..=. ..A...P.
0x0030: 19 20 c7 cd 00 00 99 17 f1 60 7a bc 1f 97 2e b7 . ...... .`z.....
0x0040: a1 18 f4 0b 5a ff 5f ac 07 71 a8 ac 54 67 3b 39 ....Z._. .q..Tg;9
0x0050: 4e 31 c5 5c 5f b5 37 ed bd 66 ee ea b1 2b 0c 26 N1.\_.7. .f...+.&
0x0060: 98 9d b8 c8 00 80 0c 57 61 87 b0 cd 08 80 00 a1 .......W a.......
```
Header data of the packet will display timestamp, source and destination info in various forms (for IPv4 protocol it is ip address, for IPv6 protocol it is IPv6 address) with MAC addresses.

Below will be displayed packet itself in form where:
- First part represents content of the packet in hexadecimal form.
- Second - ASCII form of each hexadecimal value in the first part.

### Protocols
As you have noticed, sniffer works with limitted amount of protocols. Such as:

#### IPv4
##### TCP
TCP packets are used for reliable, ordered, and error-checked
delivery of data between applications over an IP network. TCP packets operate
at the Transport layer (Layer 4) of the OSI model.

##### UDP
UDP is a transport protocol used for sending data over IP networks. It is a connectionless protocol that does not guarantee reliable delivery of data or error checking. Mostly it is used in cases when amount of data is more required than its quality (such as streaming)

##### ICMPv4
ICMPv4 is used for diagnostics and error checking only.
There is no such concept as 'port' for this type of protocol, additionaly
it operates within layer 3, while the ports are at layer 4 of OSI. For generating such traffic people usually use `ping <ip>`.

##### IGMP
IGMP protocol operates at network layer 3 of the OSI model, while ports are associated with layer 4 (transport level). IGMP is a network layer protocol used to set up multicasting on networks that use the Internet Protocol version 4 (IPv4)

#### ARP
ARP is a protocol used to map a network address
to a physical address. It has its limitations - it works only in local enviroment

#### IPv6
IPv6 is the most recent version of the Internet Protocol, designed to eventually replace IPv4.

##### NDP
NDP is a protocol in IPv6 that is used to
discover and maintain information about other nodes on the same link.NDP does not use ports, instead they use message type just like ICMPv6

##### MLD
MLD operates at the network layer (Layer 3) of the OSI model, and does not use any ports like transport layer protocols such as TCP or UDP.

##### ICMPv6
ICMPv6 is a protocol that operates at the network layer (Layer 3) of the OSI model, just like MLD. ICMPv6 messages are sent and received using IPv6 protocol, and do not use ports. ICMPv6 messages are identified by their message type field, which is part of the ICMPv6 header in the IPv6 packet.

## Testing
For testing was created separate program [packegen](https://scapy.readthedocs.io/en/latest/layers/tcp.html) in python3. It can generate packets to test IPK sniffer. It was built by me.

The goal of packegen is to test wether sniffer can filter packets correctly, display address information etc.

Additionally I have used [wireshark](https://www.wireshark.org/) in order to compare incoming packets and its content with IPK sniffer.

#### How to run test?
Packegen is using well-known library [scapy](https://scapy.readthedocs.io/en/latest/usage.html) for managing packets. 

First you will have to install scapy. You can install new environment for python
```bash
pip install virtualenv
virtualenv venv
source bin activate
pip install scapy -i https://pypi.org/simple/
```
or

```bash
pip install scapy
```

Then just run
```bash
python3 packegen.py [OPTIONS|--help]
```

Where **OPTIONS** are
- --help (prints usage info in console)
- --host (dest ip address)
- --port (dest port)
- --mode (packet type: tcp, udp, icmp4, igmp, icmp6, arp, ndp, mld)
- --timeout (interval between packets)

> **Note!**
Be aware that for IPv4 packets you will need dummy server in order to see this packets flying over your network. The server is not provided with packegen [since it's pretty easy to build](https://realpython.com/python-sockets/).

Here is the example for TCP packet
```bash
python3 packegen.py --host 123.456.7.8 --port 2023 --mode tcp
```

## Bibliography
[TCP/IP layers, IPv4 protocols. Header format](https://book.huihoo.com/iptables-tutorial/c171.htm)<br/>
[PCAP tutorial in C. How to create sniffer](https://www.tcpdump.org/pcap.html)<br/>
[IPv6 protocols and how they work](https://www.spiceworks.com/tech/networking/articles/what-is-ipv6/)<br/>
[Scapy. Python library for generating traffic](https://scapy.readthedocs.io/en/latest/layers/tcp.html)<br/>
Also was used a lot of official C header files of specific packet headers in order to understand what was inside of it.<br/>

CHANGELOG.md
# CHANGELOG

## Issues
- Program is not working if interface is not defined (-i|--interface)

## Commit history

Author: alex <italexx.cz@gmail.com>
Date:   Mon Apr 10 22:01:25 2023 +0000

    nixos basic tests + compilation

Author: turytsia <italexx.cz@gmail.com>
Date:   Mon Apr 10 19:39:32 2023 +0200

    added README documentation

Author: turytsia <turytsia.alexander@gmail.com>
Date:   Mon Apr 10 17:42:14 2023 +0200

    pre-final version

Author: turytsia <italexx.cz@gmail.com>
Date:   Sat Apr 8 00:43:50 2023 +0200

    added interface handlers

Author: Turytsa Alexander <84238200+turytsia@users.noreply.github.com>
Date:   Fri Apr 7 17:03:58 2023 +0000

    Merge pull request #2 from turytsia/feature/handle-options
    
    added pcap_findalldevs

Author: Turytsa Alexander <84238200+turytsia@users.noreply.github.com>
Date:   Fri Apr 7 17:03:52 2023 +0000

    Merge branch 'master' into feature/handle-options

Author: alex <italexx.cz@gmail.com>
Date:   Fri Apr 7 17:00:50 2023 +0000

    test

Author: turytsia <italexx.cz@gmail.com>
Date:   Fri Apr 7 18:52:38 2023 +0200

    added pcap_findalldevs

Author: Turytsa Alexander <84238200+turytsia@users.noreply.github.com>
Date:   Fri Apr 7 17:40:05 2023 +0200

    Merge pull request #1 from turytsia/feature/handle-options
    
    options handler created

Author: turytsia <italexx.cz@gmail.com>
Date:   Fri Apr 7 17:39:09 2023 +0200

    options handler created

Author: turytsia <turytsia.alexander@gmail.com>
Date:   Fri Apr 7 10:18:03 2023 +0200

    project structure created


tests/client4.py
from scapy.layers.inet import IP, UDP, ICMP
from scapy.layers.l2 import Ether, ARP
from scapy.contrib.igmp import IGMP
from scapy.all import IPv6, ICMPv6ND_NA, ICMPv6MLQuery
import socket
import time
from scapy.all import *

HOST = '172.25.3.177'
PORT = 2000

while True:
    try:
        #TCP
        # s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
        # s.setsockopt(socket.SOL_SOCKET, socket.SO_SNDBUF, 4096)
        # s.connect((HOST, PORT))
        # s.sendall(b'H')
        # Sleep for some time to keep the connection alive

        # create an IGMP packet
        # igmp_pkt = IP(dst=HOST)/IGMP(type=0x16)/Raw(load="Test IGMP packet")

        #create an ARP packet
        arp_pck = IP(dst=HOST)/ARP(pdst=HOST)/Raw(load="Test Arp packet")

        #create ndp packet
        # na_packet = IPv6(src="fe80::9f6e:fd1a:9014:4d0d", dst="fe80::215:5dff:fe5a:de1c") / \
        #     ICMPv6ND_NA(tgt="fe80::215:5dff:fe5a:de1c", R=0)
        send(arp_pck)

        time.sleep(1)

        # TCP close
        #s.close()

    except Exception as e:
        print(f'Error: {e}')


tests/client6.py
from scapy.all import *
import time

# Set the interface to use
iface = "eth0"

# Construct an ICMPv6 Neighbor Solicitation message with the target IPv6 address
ns_pkt = Ether(dst="ff:ff:ff:ff:ff:ff") / \
    IPv6(src="::", dst="fe80::215:5dff:fe5a:d592") / \
    ICMPv6ND_NS(tgt="fe80::215:5dff:fe5a:d592")

while True:
    # Send the Neighbor Solicitation message
    sendp(ns_pkt)

    time.sleep(5)


tests/server4.py
import socket

HOST = '172.25.3.177'
PORT = 2000  

with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
    s.bind((HOST, PORT))
    s.listen(1)
    print(f'Server listening on port {PORT}...')

    while True:
        # Wait for a client to connect
        conn, addr = s.accept()
        print(f'Connected by {addr}')

        with conn:

            data = conn.recv(1024)
            print(f'Received: {data.decode()}')
            # Echo the data back to the client
            conn.sendall(data)


ipk-sniffer.c
/**
 * @file ipk-sniffer.c
 * @author Oleksandr Turytsia (xturyt00)
 * @brief IPK sniffer for capturing various type of packets
 * @date 2023-04-07
 *
 * @copyright see files
 *
 */
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <stdarg.h>
#include <getopt.h>
#include <string.h>
#include <ctype.h>
#include <time.h>

 /**
  * Properly exits program with given message and errcode
  *
  * @param message Error message
  * @param ...
  */
void error(const char* message, ...) {
    va_list args;
    va_start(args, message);
    vfprintf(stderr, message, args);
    va_end(args);
    exit(EXIT_FAILURE);
}

#ifdef _WIN32

int main() {

    error("The sniffer is made for UNIX-based systems only!");

    return EXIT_FAILURE;
}

#else

#include <pcap.h>
#include <pcap/pcap.h>
#include <stdint.h>
#include <netinet/ip.h>
#include <netinet/if_ether.h>
// #include <netinet/ether.h >
#include <netinet/tcp.h>
#include <netinet/udp.h>
#include <unistd.h>
#include <netinet/icmp6.h>
#include <netinet/ip6.h>
#include <signal.h>


#define INTERFACE 'i'
#define HELP 'h'
#define FILTER_LENGTH 8

#define BUFFER_LENGTH 1024
#define BUFFER_FILTER_LENGTH 4096
#define MAC_LENGTH 18
#define TIME_LENGTH 1024
#define ETHER_SIZE 14
#define ETH_ALEN        6

#define MAX_PORT 65535

#define USAGE "./ipk-sniffer [-i interface | --interface interface] {-p port} {[--tcp|-t] [--udp|-u] [--arp] [--icmp] } {-n num}\n"

/** packet types that sniffer can watch */
enum pckfilter_t {
    TCP, UDP, ICMP4, ICMP6,
    ARP, NDP, IGMP, MLD
};

/** constants that represent filter expression*/
const char* apply_filter[] = {
    [TCP] = "tcp",
    [UDP] = "udp",
    [ICMP4] = "icmp",
    [ICMP6] = "icmp6",
    [ARP] = "arp",
    [NDP] = "(icmp6 and icmp6[0] >= 133 and icmp6 and icmp6[0] <= 136)",
    [IGMP] = "igmp",
    [MLD] = "(icmp6 and icmp6[0] >= 130 and icmp6[0] <= 132)"
};

/** Struct where all the filters are stored */
struct filter_t {
    bool pck_filter[FILTER_LENGTH];  // optional packet filters
    bool is_active;
};

/** Struct where program options are stored. */
struct opt_t {
    char interface[BUFFER_LENGTH];      // device name (interface that should be open)
    int port, npackets;                 // port and packet number
    struct filter_t filter;
};

/**  Structure that keeps track of the pointers  */
struct prog_t {
    pcap_if_t* alldevsp;    //list of all network interfaces
    pcap_t* handle;
};

/**
 *  prog is a global variable because it needs to be visible for
 *  SIGINT handler in order to free all the allocated memory
 *  and exit the program.
 */
struct prog_t prog;

/**
 * Clean up function for struct prog_t
 *
 * @param prog program that needs to clean
 */
void dump(struct prog_t* prog) {
    if (prog->alldevsp != NULL) pcap_freealldevs(prog->alldevsp);
    if (prog->handle != NULL) pcap_close(prog->handle);
}

/**
 * Checks if string 'str' contains [0-9]*
 *
 * @param str target string
 * @return Returns true if string 'str' matches [0-9]*
 */
bool is_number(char* str) {
    for (int i = 0; str[i]; i++)
        if (!isdigit(str[i])) return false;
    return true;
}

/**
 * Get the network interfaces list
 *
 * @return list of interfaces
 */
pcap_if_t* get_network_interfaces() {
    pcap_if_t* list = NULL;
    char errbuf[PCAP_ERRBUF_SIZE];

    if (pcap_findalldevs(&list, errbuf) != 0) {
        fprintf(stderr, "pcap_findalldevs(): %s\n", errbuf);
        exit(EXIT_FAILURE);
    }

    return list;
}

/**
 * Prints list of interfaces
 */
void print_network_interfaces() {
    pcap_if_t* item = get_network_interfaces();

    while (item) {
        printf("%s\n", item->name);
        item = item->next;
    }

    pcap_freealldevs(item);
    exit(EXIT_SUCCESS);
}

/**
 * Sets specific filter
 *
 * @param filter struct object
 * @param ftype  represents specific filter
 */
void set_filter(struct filter_t* filter, enum pckfilter_t ftype) {
    switch (ftype) {
    case TCP: case UDP: case ICMP4: case ICMP6:
    case ARP: case NDP: case IGMP: case MLD:
        filter->pck_filter[ftype] = true;
        filter->is_active = true;
        break;
    default:
        break;
    }
}

/**
 * Parses given arguments into a struct opt_t
 *
 * @param argc count of the arguments
 * @param argv array of arguments
 * @return structure with options
 */
struct opt_t parse_arguments(int argc, char** argv) {
    struct opt_t opt = { .port = -1, .npackets = 1, };
    int optval;                                       // option value that is being returned from getopt_long
    opterr = 0;                                       // overrides errors from getopt

    const char short_opt[] = "i:p:n:tuh";              //TODO -h / --help
    const struct option long_opt[] = {                // long options configuration
        {"interface", required_argument, 0, INTERFACE},
        {"help", no_argument, 0, HELP},
        {"tcp", no_argument, 0, TCP},
        {"udp", no_argument, 0, UDP},
        {"icmp4", no_argument, 0, ICMP4},
        {"icmp6", no_argument, 0, ICMP6},
        {"arp", no_argument, 0, ARP},
        {"ndp", no_argument, 0, NDP},
        {"igmp", no_argument, 0, IGMP},
        {"mld", no_argument, 0, MLD},
        {0,0,0,0}
    };

    while ((optval = getopt_long(argc, argv, short_opt, long_opt, NULL)) != -1) {
        switch (optval) {
        case INTERFACE:
            if (*optarg == '-')
                error("Option \'%c\' cannot be used together with other options", optval);
            else
                strcpy(opt.interface, optarg);
            break;
        case HELP:
            printf(USAGE);
            exit(EXIT_SUCCESS);
        case 'p': case 'n':
            if (!is_number(optarg))
                error("Option \'%s\' requires positive integer", optarg);

            if (optval == 'p')
                opt.port = atoi(optarg);
            else
                opt.npackets = atoi(optarg);
            break;
        case 'u': set_filter(&opt.filter, UDP); break;
        case 't': set_filter(&opt.filter, TCP); break;
        case TCP: case UDP: case ICMP4: case ICMP6:
        case ARP: case NDP: case IGMP: case MLD:
            set_filter(&opt.filter, optval); break;
        case '?':
            //TODO check if it's the only arg
            if (optopt == 'i') {
                print_network_interfaces();
            }

            if (optopt) {
                error("Option \'%c\' is invalid. Type --help or -h for help", optopt);
            }
            else {
                error("Arguments are not valid. Type --help or -h for help");
            }
            break;
        default:
            error("Unrecognized option. Type --help or -h for help");
        }
    }

    if (!opt.filter.is_active) {
        memset(opt.filter.pck_filter, true, sizeof(opt.filter.pck_filter));
    }

    return opt;
}

/**
 * converts timeval into RFC3339 time format
 *
 * @param dest destination string
 * @param tv source
 */
void timestamp2rfc3339(char* dest, const struct pcap_pkthdr* header) {
    char rfc3339[TIME_LENGTH];
    char timestamp[30];
    struct tm* time = localtime(&header->ts.tv_sec);
    strftime(timestamp, 30, "%Y-%m-%dT%H:%M:%S", time);
    sprintf(rfc3339, "%s.%03ld%+03ld:%02ld", timestamp, header->ts.tv_usec / 1000, time->tm_gmtoff / 3600,
            labs(time->tm_gmtoff % 3600) / 60);
    strcpy(dest, rfc3339);
}

/**
 * Converts bytes into its hexidecimal representation
 *
 * @param dest destination sting
 * @param bytes
 */
void bytes2hex(char* dest, uint8_t* bytes) {
    char hex[MAC_LENGTH] = { 0 };
    for (int i = 0; i < ETH_ALEN; i++) {
        char hh[4] = { 0 };
        sprintf(hh, (i < ETH_ALEN - 1 ? "%02x:" : "%02x"), bytes[i]);
        strcat(hex, hh);
    }
    strcpy(dest, hex);
}

/**
 * Prints packet in format that is specified in documentation
 *
 * @param packet contents
 * @param len
 */
void print_packet(const u_char* packet, int len) {
    int i, j, cols;
    for (i = 0; i < len; i += 16) {
        printf("\n0x%04x:", i);

        cols = i + 16;

        for (j = i; j < cols; j++) {
            if (j < len)
                printf(" %02x", packet[j]);
            else
                printf("   ");
        }
        printf(" ");
        for (j = i; cols < len ? j < cols : j < len; j++)
            printf("%c", isprint(packet[j]) ? packet[j] : '.');
    }
    printf("\n");
}

/**
 * Takes arguments and converts them into 'expression filter' string
 *
 * @param expr destination string
 * @param opt arguments
 */
void generate_filter_expr(char* expr, struct opt_t* opt) {
    char result[BUFFER_LENGTH] = { 0 };

    for (int i = 0; i < FILTER_LENGTH; i++) {
        char buff[BUFFER_FILTER_LENGTH] = { 0 };
        if (opt->filter.pck_filter[i]) {

            switch (i) {
            case TCP: case UDP:
                if (opt->port == -1) {
                    sprintf(buff, (strlen(result) > 0 ? " or %s" : "%s"), apply_filter[i]);
                }
                else {
                    sprintf(buff, (strlen(result) > 0 ? " or (%s port %d)" : "(%s port %d)"), apply_filter[i], opt->port);
                }
                break;
            case ARP: case IGMP: case ICMP4: case ICMP6:
            case MLD: case NDP:
                sprintf(buff, (strlen(result) > 0 ? " or %s" : "%s"), apply_filter[i]);
                break;
            }
            strcat(result, buff);
        }
    }
    sprintf(expr, "(%s)", result);
}

/**
 * SIGINT handler
 *
 */
void handle_signal() {
    printf("\nHERE\n");
    dump(&prog);
    exit(EXIT_SUCCESS);
}

void got_packet(u_char* args, const struct pcap_pkthdr* header,
                const u_char* packet) {
    struct ether_header* eth_header = (struct ether_header*)packet;     //packet header

    // struct pcap_pkthdr header;
    char src_dst_addr[MAC_LENGTH] = { 0 };          // mac buffer
    char timestamp[TIME_LENGTH] = { 0 };            // timestamp buffer
    char src_ip[BUFFER_LENGTH] = { 0 };             // source ip buffer
    char dest_ip[BUFFER_LENGTH] = { 0 };            // destination ip buffer

    timestamp2rfc3339(timestamp, header);
    printf("timestamp: %s\n", timestamp);
    bytes2hex(src_dst_addr, eth_header->ether_shost);
    printf("src MAC: %s\n", src_dst_addr); //TODO check this
    bytes2hex(src_dst_addr, eth_header->ether_dhost);
    printf("dst MAC: %s\n", src_dst_addr); //TODO check this
    printf("frame length: %d bytes\n", header->caplen);

    switch (ntohs(eth_header->ether_type)) {
    case ETHERTYPE_IP: {
        /** IPv4 stands for Internet Protocol version 4. It is the fourth version of the
         * Internet Protocol (IP) and is one of the core protocols of the Internet. IPv4 provides a 32-bit address space */
        

        struct ip* ip_header = (struct ip*)(packet + ETHER_SIZE);       //ipv4 header

        inet_ntop(AF_INET, &ip_header->ip_src.s_addr, src_ip, BUFFER_LENGTH);
        inet_ntop(AF_INET, &ip_header->ip_dst.s_addr, dest_ip, BUFFER_LENGTH);

        printf("src IP: %s\ndst IP: %s\n", src_ip, dest_ip);

        if (ip_header->ip_p == IPPROTO_TCP) {
            /** TCP packets are used for reliable, ordered, and error-checked
             * delivery of data between applications over an IP network. TCP packets operate
             * at the Transport layer (Layer 4) of the OSI model */

            struct tcphdr* tcp_header = (struct tcphdr*)(packet + ETHER_SIZE + sizeof(struct ip));      //tcp header
            printf("src PORT: %d\ndst PORT: %d\n", ntohs(tcp_header->th_sport), ntohs(tcp_header->th_dport));
        }
        else if (ip_header->ip_p == IPPROTO_UDP) {
            /** UDP is a transport protocol used for sending data over IP networks.
             * It is a connectionless protocol that does not guarantee reliable delivery of data or error checking.
             * Mostly it is used in cases when amount of data is more required than its quality (such as streaming) */
            struct udphdr* udp_header = (struct udphdr*)(packet + ETHER_SIZE + sizeof(struct ip));      //udp header
            printf("src PORT: %d\ndst PORT: %d\n", ntohs(udp_header->uh_sport), ntohs(udp_header->uh_dport));
        }
        else if (ip_header->ip_p == IPPROTO_ICMP) {
            /** ICMP is used for diagnostics and error checking only.
             * There is no such concept as 'port' for this type of protocol, additionaly
             * it operates within layer 3, while the ports are at layer 4 of OSI */
        }
        else if (ip_header->ip_p == IPPROTO_IGMP) {
            /** IGMP protocol operates at network layer 3 of the OSI model, while
             * ports are associated with layer 4 (transport level) */
        }
        else {
            /** Here can be handled any other IPv4 protocol. */
        }
        break;
    }
    case ETHERTYPE_ARP: {
        /** ARP is a protocol used to map a network address
         * to a physical address. It has its limitations - it works only in local enviroment */

        struct ether_arp* arp_header = (struct ether_arp*)(packet + ETHER_SIZE);        //arp header

        inet_ntop(AF_INET, &arp_header->arp_spa, src_ip, BUFFER_LENGTH);
        inet_ntop(AF_INET, &arp_header->arp_tpa, dest_ip, BUFFER_LENGTH);

        printf("src IP: %s\ndst IP: %s\n", src_ip, dest_ip);
        break;
    }
    case ETHERTYPE_IPV6: {
        /* IPv6 is the most recent version
        of the Internet Protocol, designed to eventually replace IPv4. */

        struct ip6_hdr* ip6_header = (struct ip6_hdr*)(packet + ETHER_SIZE);        //ipv6 header

        inet_ntop(AF_INET6, &ip6_header->ip6_src, src_ip, INET6_ADDRSTRLEN);
        inet_ntop(AF_INET6, &ip6_header->ip6_dst, dest_ip, INET6_ADDRSTRLEN);

        printf("src IP: %s\ndst IP: %s\n", src_ip, dest_ip);

        /** MLD operates at the network layer (Layer 3) of the OSI model,
         * and does not use any ports like transport layer protocols such as TCP or UDP */

         /** ICMPv6 is a protocol that operates
          * at the network layer (Layer 3) of the OSI model, just like MLD.
          * ICMPv6 messages are sent and received using IPv6 protocol, and do not use ports.
          * ICMPv6 messages are identified by their message type field,
          * which is part of the ICMPv6 header in the IPv6 packet.*/

          /** NDP is a protocol in IPv6 that is used to
           * discover and maintain information about other nodes on the same link.
           * NDP does not use ports, instead they use message type just like ICMPv6 */

           // NDP & MLD have too many subtypes, I decided not to add them here explicitly

        break;
    }
    default:
        /** Any other protocol over the network will be ignored since it is not part of the task. */
        break;
    }
    print_packet(packet, header->caplen);
    printf("\n");
}


int main(int argc, char** argv) {

    struct opt_t opt = parse_arguments(argc, argv); // program options

    if (optind < argc) {
        error("Option \'%s\' is not valid", argv[optind]);
    }
    
    // If [-i|--interface] is not specified, print all possible interfaces
    if (strlen(opt.interface) == 0) {
        print_network_interfaces();
    }

    char filter_exp[BUFFER_FILTER_LENGTH] = { 0 };  // filter expr buffer
    char errbuf[PCAP_ERRBUF_SIZE] = { 0 };          // error buffer

    struct bpf_program filter;
    bpf_u_int32 mask;
    bpf_u_int32 net;

    prog.alldevsp = get_network_interfaces();       // retrieve all possible interfaces

    generate_filter_expr(filter_exp, &opt);         // generate filter expression for pcap filter

    signal(SIGINT, handle_signal);                  // register signal handler for SIGINT

    if (pcap_lookupnet(opt.interface, &net, &mask, errbuf)) {
        dump(&prog);
        error("Can't get netmask for device: %s", errbuf);
    }


    prog.handle = pcap_open_live(opt.interface, BUFSIZ, 1, 1, errbuf);
    if (prog.handle == NULL) {
        dump(&prog);
        error("Unable to open device: %s", errbuf);
    }


    if (pcap_compile(prog.handle, &filter, filter_exp, 0, net) == -1) {
        dump(&prog);
        error("Unable to compile filter expression: %s", pcap_geterr(prog.handle));
    }

    if (pcap_setfilter(prog.handle, &filter) == -1) {
        dump(&prog);
        error("Unable to set filters: %s", pcap_geterr(prog.handle));
    }
    /* Capture all the packets (promiscuous mode) */

    int loop_result = pcap_loop(prog.handle, opt.npackets, got_packet, (u_char*)NULL);
    if (loop_result < 0) {
        dump(&prog);
        error("Pcap loop failed: %s", pcap_geterr(prog.handle));
    }

    dump(&prog);    //clean up

    return EXIT_SUCCESS;
}

#endif
